File: scripts/security_monitor.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
	"os/exec"
	"sort"
	"strings"
	"time"
)

const (
	version = "1.0.0"
)

// Severity levels
const (
	SeverityCritical = "CRITICAL"
	SeverityHigh     = "HIGH"
	SeverityMedium   = "MEDIUM"
	SeverityLow      = "LOW"
	SeverityInfo     = "INFO"
)

// SecurityFinding represents a single security finding
type SecurityFinding struct {
	Tool        string  `json:"tool"`
	Severity    string  `json:"severity"`
	Title       string  `json:"title"`
	Description string  `json:"description"`
	Location    string  `json:"location"`
	CWE         *string `json:"cwe,omitempty"`
	Remediation *string `json:"remediation,omitempty"`
	Timestamp   string  `json:"timestamp"`
}

// ScanReport represents the overall scan report
type ScanReport struct {
	ScanTimestamp string            `json:"scan_timestamp"`
	TotalFindings int               `json:"total_findings"`
	BySeverity    map[string]int    `json:"by_severity"`
	ByTool        map[string]int    `json:"by_tool"`
	Findings      []SecurityFinding `json:"findings"`
}

// SecurityScanner orchestrates all security scans
type SecurityScanner struct {
	findings  []SecurityFinding
	timestamp string
	workDir   string
}

// NewSecurityScanner creates a new scanner instance
func NewSecurityScanner() *SecurityScanner {
	return &SecurityScanner{
		findings:  []SecurityFinding{},
		timestamp: time.Now().UTC().Format(time.RFC3339),
		workDir:   ".",
	}
}

// RunAllScans executes all enabled security scanners
func (s *SecurityScanner) RunAllScans() error {
	log.Println("Starting comprehensive Go security scanning...")

	allPassed := true

	// Run GoSec
	if err := s.runGoSec(); err != nil {
		log.Printf("GoSec failed: %v", err)
		allPassed = false
	}

	// Run Nancy
	if err := s.runNancy(); err != nil {
		log.Printf("Nancy failed: %v", err)
		allPassed = false
	}

	// Run govulncheck
	if err := s.runGovulncheck(); err != nil {
		log.Printf("govulncheck failed: %v", err)
		allPassed = false
	}

	// Run golangci-lint with security checks
	if err := s.runGolangciLint(); err != nil {
		log.Printf("golangci-lint failed: %v", err)
		allPassed = false
	}

	// Generate reports
	if err := s.generateReports(); err != nil {
		log.Printf("Report generation failed: %v", err)
		allPassed = false
	}

	if !allPassed {
		return fmt.Errorf("one or more security scans failed")
	}

	return nil
}

// runGoSec executes the GoSec security scanner
func (s *SecurityScanner) runGoSec() error {
	log.Println("Running GoSec security scanner...")

	cmd := exec.Command("gosec",
		"-conf", ".security/.gosec.yml",
		"-fmt", "json",
		"-out", "gosec-report.json",
		"./...")
	cmd.Dir = s.workDir

	var stderr bytes.Buffer
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		// GoSec returns error if findings exist
		if stderr.Len() > 0 {
			log.Printf("GoSec stderr: %s", stderr.String())
		}
	}

	// Parse results
	data, err := os.ReadFile("gosec-report.json")
	if err != nil {
		log.Printf("Could not read gosec-report.json: %v", err)
		return nil // Don't fail if file doesn't exist
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return fmt.Errorf("failed to parse gosec report: %w", err)
	}

	issues, ok := result["Issues"].([]interface{})
	if ok {
		for _, issue := range issues {
			if issueMap, ok := issue.(map[string]interface{}); ok {
				severity := "MEDIUM"
				if sev, ok := issueMap["severity"].(string); ok {
					severity = sev
				}

				location := ""
				if file, ok := issueMap["file"].(string); ok {
					location = file
					if line, ok := issueMap["line"].(float64); ok {
						location = fmt.Sprintf("%s:%d", file, int(line))
					}
				}

				finding := SecurityFinding{
					Tool:        "gosec",
					Severity:    severity,
					Title:       "Security Issue",
					Description: getString(issueMap, "details"),
					Location:    location,
					CWE:         getStringPtr(issueMap, "cwe"),
					Timestamp:   s.timestamp,
				}
				s.findings = append(s.findings, finding)
			}
		}
	}

	log.Println("GoSec scan completed")
	return nil
}

// runNancy executes the Nancy dependency checker
func (s *SecurityScanner) runNancy() error {
	log.Println("Running Nancy dependency scanner...")

	cmd := exec.Command("sh", "-c", "go list -json -m all | nancy sleuth -o json > nancy-report.json")
	cmd.Dir = s.workDir

	if err := cmd.Run(); err != nil {
		log.Printf("Nancy execution failed: %v", err)
		// Continue even if nancy fails
	}

	data, err := os.ReadFile("nancy-report.json")
	if err != nil {
		log.Printf("Could not read nancy-report.json: %v", err)
		return nil
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return fmt.Errorf("failed to parse nancy report: %w", err)
	}

	// Parse Nancy results (format varies)
	if vulns, ok := result["vulnerabilities"].([]interface{}); ok {
		for _, vuln := range vulns {
			if vulnMap, ok := vuln.(map[string]interface{}); ok {
				finding := SecurityFinding{
					Tool:        "nancy",
					Severity:    "HIGH",
					Title:       getString(vulnMap, "package"),
					Description: getString(vulnMap, "vulnerability"),
					Location:    "dependencies",
					Timestamp:   s.timestamp,
				}
				s.findings = append(s.findings, finding)
			}
		}
	}

	log.Println("Nancy scan completed")
	return nil
}

// runGovulncheck executes the Go vulnerability database checker
func (s *SecurityScanner) runGovulncheck() error {
	log.Println("Running govulncheck...")

	cmd := exec.Command("govulncheck", "-json", "./...")
	cmd.Dir = s.workDir

	output, err := cmd.Output()
	if err != nil {
		// govulncheck returns error if vulnerabilities found
		log.Printf("govulncheck found issues")
	}

	// Parse line-delimited JSON output
	for _, line := range strings.Split(string(output), "\n") {
		if strings.TrimSpace(line) == "" {
			continue
		}

		var result map[string]interface{}
		if err := json.Unmarshal([]byte(line), &result); err != nil {
			continue
		}

		if vulnType, ok := result["type"].(string); ok && vulnType == "Vuln" {
			finding := SecurityFinding{
				Tool:        "govulncheck",
				Severity:    "HIGH",
				Title:       getString(result, "vuln_id"),
				Description: getString(result, "summary"),
				Location:    getString(result, "package"),
				Timestamp:   s.timestamp,
			}
			s.findings = append(s.findings, finding)
		}
	}

	log.Println("govulncheck completed")
	return nil
}

// runGolangciLint executes golangci-lint with security checks
func (s *SecurityScanner) runGolangciLint() error {
	log.Println("Running golangci-lint with security checks...")

	cmd := exec.Command("golangci-lint",
		"run",
		"--config", ".golangci.yml",
		"--out-format", "json",
		"-o", "golangci-report.json",
		"./...")
	cmd.Dir = s.workDir

	if err := cmd.Run(); err != nil {
		// golangci-lint returns error if issues found
		log.Printf("golangci-lint found issues")
	}

	data, err := os.ReadFile("golangci-report.json")
	if err != nil {
		log.Printf("Could not read golangci-report.json: %v", err)
		return nil
	}

	var result map[string]interface{}
	if err := json.Unmarshal(data, &result); err != nil {
		return fmt.Errorf("failed to parse golangci report: %w", err)
	}

	issues, ok := result["Issues"].([]interface{})
	if ok {
		for _, issue := range issues {
			if issueMap, ok := issue.(map[string]interface{}); ok {
				// Only include security-related checks
				linter := getString(issueMap, "from_linter")
				if !isSecurityLinter(linter) {
					continue
				}

				location := fmt.Sprintf("%s:%v", getString(issueMap, "pos"), "")
				finding := SecurityFinding{
					Tool:        "golangci-lint",
					Severity:    "MEDIUM",
					Title:       linter,
					Description: getString(issueMap, "text"),
					Location:    location,
					Timestamp:   s.timestamp,
				}
				s.findings = append(s.findings, finding)
			}
		}
	}

	log.Println("golangci-lint scan completed")
	return nil
}

// isSecurityLinter checks if the linter is security-related
func isSecurityLinter(linter string) bool {
	securityLinters := map[string]bool{
		"gosec":       true,
		"staticcheck": true,
		"gas":         true,
	}
	return securityLinters[linter]
}

// generateReports generates various report formats
func (s *SecurityScanner) generateReports() error {
	log.Println("Generating security reports...")

	// Count by severity
	bySeverity := make(map[string]int)
	for _, finding := range s.findings {
		bySeverity[finding.Severity]++
	}

	// Count by tool
	byTool := make(map[string]int)
	for _, finding := range s.findings {
		byTool[finding.Tool]++
	}

	report := ScanReport{
		ScanTimestamp: s.timestamp,
		TotalFindings: len(s.findings),
		BySeverity:    bySeverity,
		ByTool:        byTool,
		Findings:      s.findings,
	}

	// Generate JSON report
	if err := s.generateJSONReport(report); err != nil {
		return fmt.Errorf("failed to generate JSON report: %w", err)
	}

	// Generate Markdown report
	if err := s.generateMarkdownReport(report); err != nil {
		return fmt.Errorf("failed to generate Markdown report: %w", err)
	}

	// Generate SARIF report for 
	if err := s.generateSARIFReport(report); err != nil {
		return fmt.Errorf("failed to generate SARIF report: %w", err)
	}

	return nil
}

// generateJSONReport exports findings as JSON
func (s *SecurityScanner) generateJSONReport(report ScanReport) error {
	data, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile("security-report.json", data, 0644)
}

// generateMarkdownReport generates a human-readable markdown report
func (s *SecurityScanner) generateMarkdownReport(report ScanReport) error {
	var buf bytes.Buffer

	buf.WriteString("# Go Security Scan Report\n\n")
	buf.WriteString(fmt.Sprintf("**Scan Timestamp:** %s\n\n", report.ScanTimestamp))
	buf.WriteString(fmt.Sprintf("**Total Findings:** %d\n\n", report.TotalFindings))

	// Summary by severity
	buf.WriteString("## Summary by Severity\n\n")
	for _, severity := range []string{SeverityCritical, SeverityHigh, SeverityMedium, SeverityLow, SeverityInfo} {
		if count, ok := report.BySeverity[severity]; ok && count > 0 {
			buf.WriteString(fmt.Sprintf("- **%s:** %d\n", severity, count))
		}
	}

	// Summary by tool
	buf.WriteString("\n## Summary by Tool\n\n")
	tools := make([]string, 0, len(report.ByTool))
	for tool := range report.ByTool {
		tools = append(tools, tool)
	}
	sort.Strings(tools)

	for _, tool := range tools {
		buf.WriteString(fmt.Sprintf("- **%s:** %d\n", tool, report.ByTool[tool]))
	}

	// Detailed findings
	if len(report.Findings) > 0 {
		buf.WriteString("\n## Detailed Findings\n\n")
		for _, finding := range report.Findings {
			buf.WriteString(fmt.Sprintf("### %s\n\n", finding.Title))
			buf.WriteString(fmt.Sprintf("- **Tool:** %s\n", finding.Tool))
			buf.WriteString(fmt.Sprintf("- **Severity:** %s\n", finding.Severity))
			buf.WriteString(fmt.Sprintf("- **Location:** %s\n", finding.Location))
			if finding.Description != "" {
				buf.WriteString(fmt.Sprintf("- **Description:** %s\n", finding.Description))
			}
			if finding.CWE != nil && *finding.CWE != "" {
				buf.WriteString(fmt.Sprintf("- **CWE:** %s\n", *finding.CWE))
			}
			if finding.Remediation != nil && *finding.Remediation != "" {
				buf.WriteString(fmt.Sprintf("- **Remediation:** %s\n", *finding.Remediation))
			}
			buf.WriteString("\n")
		}
	}

	return os.WriteFile("security-report.md", buf.Bytes(), 0644)
}

// generateSARIFReport generates SARIF format for  integration
func (s *SecurityScanner) generateSARIFReport(report ScanReport) error {
	sarif := map[string]interface{}{
		"version": "2.1.0",
		"runs": []map[string]interface{}{
			{
				"tool": map[string]interface{}{
					"driver": map[string]interface{}{
						"name":           "PAW Security Scanner",
						"version":        version,
						"informationUri": "https://github.com/paw-org/paw",
					},
				},
				"results": s.buildSARIFResults(),
			},
		},
	}

	data, err := json.MarshalIndent(sarif, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile("security-report.sarif", data, 0644)
}

// buildSARIFResults converts findings to SARIF results format
func (s *SecurityScanner) buildSARIFResults() []map[string]interface{} {
	var results []map[string]interface{}

	for _, finding := range s.findings {
		parts := strings.Split(finding.Location, ":")
		file := parts[0]
		line := 1
		if len(parts) > 1 {
			fmt.Sscanf(parts[1], "%d", &line)
		}

		result := map[string]interface{}{
			"ruleId": finding.Tool,
			"level":  strings.ToLower(finding.Severity),
			"message": map[string]interface{}{
				"text": finding.Title,
			},
			"locations": []map[string]interface{}{
				{
					"physicalLocation": map[string]interface{}{
						"artifactLocation": map[string]string{
							"uri": file,
						},
						"region": map[string]int{
							"startLine": line,
						},
					},
				},
			},
		}

		results = append(results, result)
	}

	return results
}

// Helper functions

func getString(m map[string]interface{}, key string) string {
	if val, ok := m[key]; ok {
		if str, ok := val.(string); ok {
			return str
		}
	}
	return ""
}

func getStringPtr(m map[string]interface{}, key string) *string {
	val := getString(m, key)
	if val != "" {
		return &val
	}
	return nil
}

func main() {
	flag.Parse()

	scanner := NewSecurityScanner()

	if err := scanner.RunAllScans(); err != nil {
		log.Printf("Security scanning failed: %v", err)
		os.Exit(1)
	}

	log.Println("Security scanning completed successfully")
	os.Exit(0)
}

File: scripts/security_monitor.py
#!/usr/bin/env python3
"""
PAW Blockchain - Continuous Security Monitoring System (Python Component)

Automated security scanning for Python code, including:
- Bandit: Security issue detection
- Safety: Dependency vulnerability checking
- pip-audit: Package vulnerability auditing
- Semgrep: Static analysis and SAST
"""

import json
import os
import sys
import subprocess
import logging
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict
from enum import Enum
import yaml

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class Severity(Enum):
    """Security finding severity levels"""
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"
    INFO = "INFO"

    def __lt__(self, other):
        severity_order = {
            "CRITICAL": 0,
            "HIGH": 1,
            "MEDIUM": 2,
            "LOW": 3,
            "INFO": 4
        }
        return severity_order[self.value] < severity_order[other.value]


@dataclass
class SecurityFinding:
    """Represents a security finding"""
    tool: str
    severity: str
    title: str
    description: str
    location: str
    cwe: Optional[str] = None
    remediation: Optional[str] = None
    timestamp: Optional[str] = None

    def to_dict(self) -> Dict:
        return asdict(self)


class SecurityScanner:
    """Main security scanning orchestrator"""

    def __init__(self, config_path: str = ".security/config.yml"):
        self.config_path = config_path
        self.config = self._load_config()
        self.findings: List[SecurityFinding] = []
        self.scan_timestamp = datetime.now().isoformat()
        self.project_root = Path(__file__).parent.parent

    def _load_config(self) -> Dict[str, Any]:
        """Load security configuration from YAML"""
        config_file = Path(self.config_path)
        if not config_file.exists():
            logger.warning(f"Config file not found: {self.config_path}, using defaults")
            return {}

        try:
            with open(config_file, 'r') as f:
                return yaml.safe_load(f) or {}
        except Exception as e:
            logger.error(f"Failed to load config: {e}")
            return {}

    def run_all_scans(self) -> bool:
        """Run all enabled security scanners"""
        logger.info("Starting comprehensive Python security scanning...")

        python_config = self.config.get('python_security', {})
        tools = python_config.get('tools', {})

        all_passed = True

        if tools.get('bandit', {}).get('enabled', True):
            all_passed &= self._run_bandit(tools.get('bandit', {}))

        if tools.get('safety', {}).get('enabled', True):
            all_passed &= self._run_safety(tools.get('safety', {}))

        if tools.get('pip_audit', {}).get('enabled', True):
            all_passed &= self._run_pip_audit(tools.get('pip_audit', {}))

        if tools.get('semgrep', {}).get('enabled', True):
            all_passed &= self._run_semgrep(tools.get('semgrep', {}))

        # Generate report
        self._generate_report()

        return all_passed

    def _run_bandit(self, config: Dict) -> bool:
        """Run Bandit security scanner"""
        logger.info("Running Bandit security scanner...")
        try:
            cmd = [
                'bandit',
                '-r', 'src/',
                '-f', 'json',
                '-o', 'bandit-report.json'
            ]

            severity = config.get('severity', 'medium')
            if severity:
                cmd.extend(['-ll'] if severity == 'low' else [])

            result = subprocess.run(cmd, capture_output=True, text=True)

            # Parse results
            if os.path.exists('bandit-report.json'):
                with open('bandit-report.json', 'r') as f:
                    data = json.load(f)
                    for issue in data.get('results', []):
                        finding = SecurityFinding(
                            tool='bandit',
                            severity=issue.get('severity', 'MEDIUM'),
                            title=issue.get('test_name', 'Unknown'),
                            description=issue.get('issue_text', ''),
                            location=f"{issue.get('filename')}:{issue.get('line_number')}",
                            cwe=f"CWE-{issue.get('test_id', 'Unknown')}",
                            timestamp=self.scan_timestamp
                        )
                        self.findings.append(finding)

            # Check if we should fail
            fail_on_issues = config.get('fail_on_issues', True)
            has_findings = len(self.findings) > 0

            if fail_on_issues and has_findings:
                logger.warning(f"Bandit found {len(self.findings)} security issues")
                return False

            logger.info("Bandit scan completed successfully")
            return True

        except FileNotFoundError:
            logger.error("Bandit not installed. Install with: pip install bandit")
            return False
        except Exception as e:
            logger.error(f"Bandit scan failed: {e}")
            return False

    def _run_safety(self, config: Dict) -> bool:
        """Run Safety dependency vulnerability checker"""
        logger.info("Running Safety vulnerability check...")
        try:
            cmd = ['safety', 'check', '--json']

            result = subprocess.run(cmd, capture_output=True, text=True)

            if result.returncode != 0:
                try:
                    data = json.loads(result.stdout)
                    for vuln in data:
                        if isinstance(vuln, dict):
                            finding = SecurityFinding(
                                tool='safety',
                                severity='HIGH',  # Safety reports are usually high priority
                                title=vuln.get('package', 'Unknown'),
                                description=vuln.get('vulnerability', ''),
                                location='dependencies',
                                cwe=None,
                                timestamp=self.scan_timestamp
                            )
                            self.findings.append(finding)
                except json.JSONDecodeError:
                    pass

            fail_on_vuln = config.get('fail_on_vulnerability', True)
            if fail_on_vuln and len(self.findings) > 0:
                logger.warning("Safety found vulnerabilities in dependencies")
                return False

            logger.info("Safety check completed")
            return True

        except FileNotFoundError:
            logger.error("Safety not installed. Install with: pip install safety")
            return False
        except Exception as e:
            logger.error(f"Safety check failed: {e}")
            return False

    def _run_pip_audit(self, config: Dict) -> bool:
        """Run pip-audit for package vulnerability scanning"""
        logger.info("Running pip-audit...")
        try:
            cmd = ['pip-audit', '--format', 'json']

            result = subprocess.run(cmd, capture_output=True, text=True)

            if result.returncode != 0:
                try:
                    data = json.loads(result.stdout)
                    for vuln in data.get('vulnerabilities', []):
                        finding = SecurityFinding(
                            tool='pip-audit',
                            severity='HIGH',
                            title=vuln.get('name', 'Unknown'),
                            description=vuln.get('description', ''),
                            location='dependencies',
                            cwe=None,
                            remediation=vuln.get('fix_version'),
                            timestamp=self.scan_timestamp
                        )
                        self.findings.append(finding)
                except json.JSONDecodeError:
                    pass

            fail_on_vuln = config.get('fail_on_vulnerability', True)
            if fail_on_vuln and len(self.findings) > 0:
                logger.warning("pip-audit found vulnerabilities")
                return False

            logger.info("pip-audit completed")
            return True

        except FileNotFoundError:
            logger.error("pip-audit not installed. Install with: pip install pip-audit")
            return False
        except Exception as e:
            logger.error(f"pip-audit failed: {e}")
            return False

    def _run_semgrep(self, config: Dict) -> bool:
        """Run Semgrep static analysis"""
        logger.info("Running Semgrep analysis...")
        try:
            cmd = [
                'semgrep',
                '--config', config.get('config', 'p/security-audit'),
                '--json',
                '--output', 'semgrep-report.json',
                'src/'
            ]

            result = subprocess.run(cmd, capture_output=True, text=True)

            if os.path.exists('semgrep-report.json'):
                with open('semgrep-report.json', 'r') as f:
                    data = json.load(f)
                    for result_item in data.get('results', []):
                        severity = result_item.get('extra', {}).get('severity', 'MEDIUM')
                        finding = SecurityFinding(
                            tool='semgrep',
                            severity=severity,
                            title=result_item.get('check_id', 'Unknown'),
                            description=result_item.get('extra', {}).get('message', ''),
                            location=f"{result_item.get('path')}:{result_item.get('start', {}).get('line')}",
                            cwe=None,
                            timestamp=self.scan_timestamp
                        )
                        self.findings.append(finding)

            fail_on_findings = config.get('fail_on_findings', True)
            critical_findings = [f for f in self.findings if f.tool == 'semgrep' and f.severity in ['CRITICAL', 'HIGH']]

            if fail_on_findings and critical_findings:
                logger.warning(f"Semgrep found {len(critical_findings)} critical/high findings")
                return False

            logger.info("Semgrep analysis completed")
            return True

        except FileNotFoundError:
            logger.error("Semgrep not installed. Install with: pip install semgrep")
            return False
        except Exception as e:
            logger.error(f"Semgrep analysis failed: {e}")
            return False

    def _generate_report(self):
        """Generate security report"""
        report_data = {
            'scan_timestamp': self.scan_timestamp,
            'total_findings': len(self.findings),
            'by_severity': self._count_by_severity(),
            'by_tool': self._count_by_tool(),
            'findings': [f.to_dict() for f in self.findings]
        }

        # Save JSON report
        with open('security-report.json', 'w') as f:
            json.dump(report_data, f, indent=2)

        # Generate markdown report
        self._generate_markdown_report(report_data)

        logger.info(f"Security report generated with {len(self.findings)} findings")

    def _count_by_severity(self) -> Dict[str, int]:
        """Count findings by severity"""
        counts = {s.value: 0 for s in Severity}
        for finding in self.findings:
            if finding.severity in counts:
                counts[finding.severity] += 1
        return counts

    def _count_by_tool(self) -> Dict[str, int]:
        """Count findings by tool"""
        counts = {}
        for finding in self.findings:
            counts[finding.tool] = counts.get(finding.tool, 0) + 1
        return counts

    def _generate_markdown_report(self, data: Dict):
        """Generate markdown security report"""
        lines = [
            "# Python Security Scan Report",
            f"\n**Scan Timestamp:** {data['scan_timestamp']}",
            f"\n**Total Findings:** {data['total_findings']}\n"
        ]

        # Summary by severity
        lines.append("## Summary by Severity\n")
        for severity, count in data['by_severity'].items():
            if count > 0:
                lines.append(f"- **{severity}:** {count}")

        # Summary by tool
        lines.append("\n## Summary by Tool\n")
        for tool, count in data['by_tool'].items():
            lines.append(f"- **{tool}:** {count}")

        # Detailed findings
        if data['findings']:
            lines.append("\n## Detailed Findings\n")
            for finding in data['findings']:
                lines.append(f"### {finding['title']}")
                lines.append(f"- **Tool:** {finding['tool']}")
                lines.append(f"- **Severity:** {finding['severity']}")
                lines.append(f"- **Location:** {finding['location']}")
                if finding.get('description'):
                    lines.append(f"- **Description:** {finding['description']}")
                if finding.get('remediation'):
                    lines.append(f"- **Remediation:** {finding['remediation']}")
                lines.append("")

        with open('security-report.md', 'w') as f:
            f.write('\n'.join(lines))

    def export_sarif(self):
        """Export findings in SARIF format for """
        sarif = {
            "version": "2.1.0",
            "runs": [{
                "tool": {
                    "driver": {
                        "name": "PAW Security Scanner",
                        "version": "1.0.0",
                        "informationUri": "https://example.com/paw-org/paw"
                    }
                },
                "results": []
            }]
        }

        for finding in self.findings:
            result = {
                "ruleId": finding.tool,
                "level": finding.severity.lower(),
                "message": {
                    "text": finding.title
                },
                "locations": [{
                    "physicalLocation": {
                        "artifactLocation": {
                            "uri": finding.location.split(':')[0]
                        },
                        "region": {
                            "startLine": int(finding.location.split(':')[1]) if ':' in finding.location else 1
                        }
                    }
                }]
            }
            sarif["runs"][0]["results"].append(result)

        with open('security-report.sarif', 'w') as f:
            json.dump(sarif, f, indent=2)

        logger.info("SARIF report generated")


def main():
    """Main entry point"""
    scanner = SecurityScanner()
    success = scanner.run_all_scans()
    scanner.export_sarif()

    # Exit with appropriate code
    sys.exit(0 if success else 1)


if __name__ == '__main__':
    main()
